compiler/rustc_errors/src/lib.rs

```diff
@@ -795,20 +795,16 @@ impl HandlerInner {
 
     fn emit_diagnostic(&mut self, diagnostic: &Diagnostic) {
         if diagnostic.cancelled() {
+            if !self.flags.can_emit_warnings && diagnostic.has_future_breakage() {
+                (*TRACK_DIAGNOSTICS)(diagnostic);
+            }
```

Comment:
This seems to change the behavior of `can_emit_warnings`. In particular, if there is a non-lint-based warning, it is now allowed to be emitted (that is, something that creates a warning diagnostic without going through `struct_warn`).  Those are very rare.  This seems to hobble `can_emit_warnings` such that it only impacts non-lints that use `struct_warn`, which is also very rare.

This also doesn't seem to be correct for TRACK_DIAGNOSTICS.  I believe it should only be called if the diagnostic is pushed onto `future_breakage_diagnostics` below. It is not clear to me how important it is that TRACK_DIAGNOSTICS gets called. It appears to have something to do with incremental, but what happens if it is not called?

Also, the old code seems broken to me. I don't see how it is possible for a diagnostic to have a level of Warning and `can_emit_warnings` is false (I believe the level will always be Allow with `--cap-lints=allow`).  It seems like that call to TRACK_DIAGNOSTICS would never happen.

In general, the problem seems that `emit_diagnostic` does not have enough information to know if this was a force-warns diagnostic.  Maybe DiagnosticId::Lint could add a field to indicate it was force-warn (similar to the has_future_breakage field)?

Reply:
I like the suggestion of adding this to `DiagnosticId::Lint`, but I'm going to wait for https://github.com/rust-lang/rust/pull/86009 to land before addressing this since I think how we do this will change slightly once that lands.